manual/src/cmds/intf-c.etex

```diff
@@ -580,8 +580,21 @@ that is, as pointers to blocks tagged~0.  They are accessed with the
 Arrays of floating-point numbers (type "float array")
 have a special, unboxed, more efficient representation.
 These arrays are represented by pointers to blocks with tag
-"Double_array_tag".  They should be accessed with the "Double_field"
-and "Store_double_field" macros.
+"Double_array_tag".  They should be accessed with the "Double_array_field"
+and "Store_double_array_field" macros.
+
+The special representation of "float array" necessitates an extra runtime check
+in most polymorphic array operations. To avoid this penalty, a dedicated type
+"floatarray" is also available which enjoys the special representation
+unconditionally. Values of type "floatarray", as well as records consisting
+solely of floats, should be accessed using the "Double_flat_field" and
+"Store_double_flat_field" macros.
+
+While currently "float array" and "floatarray" enjoy the same unboxed
+representation, it is possible to revert "float array" to its natural unboxed
+representation by passing the "--disable-flat-float-array" flag to the
+"configure" script (in this case, values of type "float array" will be
+tagged~0). The above macros will continue to work in either case.
```

Comment:
There are two things I am not fond of with your proposal:
- The second proposal suggests that `float array` values are unboxed, and that `Double_array_field` is used for unboxed arrays; it is only in the fourth paragraph that we learn that in fact they may be represented differently. It is easy to read 2 in isolation and come to the wrong conclusion.
- The documentation of how to access each type is intertwined for justifications for having this or that option, which make the whole thing harder to read, and also increases the chances that other people object to the description (the consensus on default choice of array representations is muddy).

My proposal: start by documenting things that have an unambiguous representation, then mention `float array` as having either of them, without too much discussion of reason for having the three versions. (If we want to document that, which sounds reasonable, I propose to delay it to another PR.)


Concrete proposal:

```
Arrays of integers and pointers are represented like tuples and records,
that is, as pointers to blocks tagged~0.  They are accessed with the
"Field" macro for reading and the "caml_modify" function for writing.

Values of type "floatarray" (as manipulated by the
"Array.Floatarray" module), as well as records whose declaration
contains only float fields, use an efficient "flat" or unboxed
representation: raw blocks whose content are the raw double values
directly, which are not valid OCaml values. They should be accessed
using the "Double_flat_field" and "Store_double_flat_field" macros.

Finally, arrays of type "float array" use one or the other
representation depending on the compiler and runtime
configuration. They are currently "flat" by default, and "boxed" when
the "--disable-flat-float-array" configure flag is used. They should
be accessed using the "Double_array_field" and
"Store_double_array_field" macros, that will work correctly under both
configurations.
```

I am left with a question: if I have a value that I know has some `t array` type for an abstract type `t` which may or may not be `float`, what macros am I supposed to use? Do any of the three macros mentioned here work?

Reply:
Thanks @gasche for your suggestion! I pushed something very close to your suggestion (mostly I avoided the "flat" terminology which I don't find very clear, and used only "boxed/unboxed" instead; and also tweaked a few words here and there).

> I am left with a question: if I have a value that I know has some `t array` type for an abstract type `t` which may or may not be `float`, what macros am I supposed to use? Do any of the three macros mentioned here work?

No. You need to check the tag before deciding what to use; essentially what the `caml_array_set` and `caml_array_get` primitives do.